_posts/2023-11-08-device-reliability-metrics.md

```diff
@@ -0,0 +1,365 @@
+
+To calculate crash free devices, you need three things: a chosen time window to aggregate the data by (daily, weekly, or monthly), the number of devices that experienced a crash in that window, and the total number of devices active.
+
+<p align="center">
+  <img width="650" src="{% img_url device-reliability-metrics/crash-free-devices-formula.png %}" alt="Formula for how to calculate crash free devices" />
+</p>
+
+Each device will send data points about when it crashed, and the heavy lifting is done in the data warehouse. One can use the following SQL query to calculate crash-free devices in the last 7 days.
+
+```sql
+SELECT
+    count(DISTINCT device_id),
+FROM
+    device_crashes
+WHERE
+    crashed_at > NOW() - INTERVAL '7 days';
+```
+
+Below is an example of what the data from the device might look like and the resulting stat.
+
+<p align="center">
+  <img width="600" src="{% img_url device-reliability-metrics/crash-free-devices.png %}" alt="Diagram of how crash free devices is collected" />
+</p>
+
+This metric will show your firmware’s true reliability score, as it only takes one crash on each device to get to 0%! The other metrics are not this sensitive, so only show this metric to your manager and CTO if you are committed to improving it.
+
+#### Collection
+
+Upon boot, send whether the device reset due to a crash and a time stamp to help place the crash into the correct day bucket.
+
+#### Summary
+
+Below is the standard comparison chart, but there is one final note about crash free devices that I need to make. This metric does not work well to compare device reliability _between firmware releases._ To determine whether your firmware update is better or worse than the one it’s replacing, use a different metric.
+
+| Criteria                                                                        | Rating | Notes                                                                         |
+| ------------------------------------------------------------------------------- | ------ | ----------------------------------------------------------------------------- |
+| Can assess the reliability of devices & software quickly after firmware updates | ⚠️     | To get crash free devices over N days, you need to wait N days, so it’s slow. |
+| Handles expected vs unexpected reboots                                          | ✅     | Expected reboots performed by the user are properly ignored.                  |
+| Not susceptible to a small subset of misbehaving devices skewing the metric     | ✅     | 1 crash is the same as 100 over N days.                                       |
+| Works well with session-based devices                                           | ✅     | No issues.                                                                    |
+
+<br>
+
+## Metrics to Give the Boss
+
+If I were Head of Hardware or Firmware at an IoT company, I would want the following information in real-time:
```

Comment:
I'm struggling to think of a natural way to fit it in, but I think it would be nice to have a link to your battery post somewhere in here.